design/design-generic-references.md

```diff
@@ -0,0 +1,319 @@
+# Generic Resource References
+
+* Owner: Predrag Knezevic (@pedjak)
+* Reviewers: Crossplane Maintainers
+* Status: Draft
+
+## Background
+
+Crossplane providers provision and monitor external resources in third party
+cloud providers, following the declared specification in the corresponding managed
+resources. Very often, creation of these objects cannot happen in parallel, i.e. some
+fields can be set only after a given dependency is created.
+
+[Cross resource referencing] solves this problem for majority of cases by
+following [Kubernetes API object reference] convention:
+
+```yaml
+spec:
+  fooRef:
+    name: bar
+```
+
+Field `spec.fooRef` contains the name of the resource where the value for `spec.foo` is to
+be found. The kind of referred resource and the field containing the value used later for
+setting `spec.foo` is determined during compile time.
+
+Although the used convention covers the majority of use cases, it lacks the
+following properties:
+
+* The referred object kind and the source field cannot be changed at 
+  runtime for a given object
+* The referencer field may not always target a single referred kind. For
+  example, a URL field in Route53 could get populated by a URL of an S3 bucket
+  or access endpoint of an RDS Instance
+* The referencer field may not always target a single field in the referred
+  object
+* The value that is about to be assigned to the field require some sort of
+  transformation
+
+Some of the above issues can be solved by defining a composition. However, there
+are cases where one would like to use and refer to a resource not bound by a 
+composite instance.
```

Comment:
Compositions today mix specification, relationships and orchestration:
1. **specification**: specify how objects look like
2. **relationships**: Patch&Transformations (P&T) establish field level dependencies
3. **orchestration**: P&T is used to define order, both by (a) implicit field dependencies of required fields and (b) by hacks to intentionally break objects or pause objects before some condition is true.

In that light, I would consider the goal of this proposal to promote (2) to the object system, and primarily leave compositions for (1) and (3). This bigger "philosophical" context must be clearer in this doc, maybe in some motivation section.

There is also a connection between (2) and (3a). An asynchronous controller cannot solve the required fields problem, i.e. late initialization. This means that we encourage late initialization more and more. Late initialization has the problem of more states to reason about. Resources will not be completely or correctly specified fromt he beginning, i.e. the creation of the object. This obviously has consequences: less invariants/consistency, less type-safety. 

Is that ^^ a higher level goal? Isn't being more dynamic, less consistent, less type-safety a step into the wrong direction? Thinking about DevEx and "Crossplane is complex".